cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java

```diff
@@ -218,6 +236,43 @@ protected void run() {
       return;
     }
 
+    if (instanceId == 0) {
+      // todo iterate migration resolution order list and schedule execution based on data availability.
+
+      // transfer aggregates metrics data at the beginning, in a single run to avoid conflicts with new metrics written
+      // as metrics table is non-transactional. having conflict will provide incorrect results.
+      // it is okay to run this without a limit on records, as the total records in aggregates table will be equal
+      // to the total unique metrics in the system.
+      // we will schedule the transfer of metrics data for time resolution tables
+      // in a single thread executor with limit on records to transfer.
+      // since the smaller resolution tables have timestamp information in the rowkey, they wont have conflict for most
+      // of their data, except for the few metrics emitted in the last hour/minute.
+
+      if (!skipMigration) {
+        int maxRecordsToScan = 1000;
+        LinkedHashMap<Integer, Integer> resolutionToMaxRecordsMap = new LinkedHashMap<>();
+        // migrate aggregates resolution table without limit on max records scan
+        resolutionToMaxRecordsMap.put(Integer.MAX_VALUE, Integer.MAX_VALUE);
+        // migrate 1hr resolution table, then 1min resolution table,
+        // limit 1000 records to scan/transfer in a run.
+        resolutionToMaxRecordsMap.put(3600, maxRecordsToScan);
+        resolutionToMaxRecordsMap.put(60, maxRecordsToScan);
+
+        ScheduledExecutorService dataMigratorExecutor =
+          Executors.newSingleThreadScheduledExecutor(Threads.createDaemonThreadFactory("metrics-data-migrator"));
```

Comment:
who will shutdown these executors if metrics processor is shutdown? (before the migration is complete)